Code changes for Collecting Device Logs - Killing the Emulator Process - Setting Permissions in iOS

sarangan12 opened the pull request:
The following functionalities have been added to Paramedic code.

1. To collect device logs
2. To kill the emulator processes 
3. To set permissions to contacts for the test app

@omefire @dblotsky @riknoll @nikhilkh @sgrebnov @vladimir-kotikov Can you please review and merge this PR?

riknoll commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+
+		//Check if the simFolder has TCC.db file in it
+		if(self.doesFileExist(destinationTCCFile)) {
+			var sqlite3Command_firstPart  = 'sqlite3 ' + destinationTCCFile;
```
Why split this variable into three parts like this?

riknoll commented on lib/paramedic.js:
```diff
@@ -17,22 +17,30 @@
+var Q       = require('q');
+var tmp     = require('tmp');
+var path    = require('path');
+var Q       = require('q');
```
Q is being required twice

riknoll commented:
Can paramedicPermissions be renamed to make it clear that it's only for iOS? The other naming convention seems like it applies to the generic files.

dblotsky commented on lib/ParamedicConfig.js:
```diff
@@ -37,7 +37,12 @@ ParamedicConfig.parseFromArguments = function (argv) {
+        logMins:           !!argv.logMins? argv.logMins: undefined,
+        appName:           !!argv.appName? argv.appName: undefined,
+        simulatorsFolder:  !!argv.simulatorsFolder? argv.simulatorsFolder: undefined,
+        tccDbPath:         !!argv.tccDbPath? argv.tccDbPath: undefined
```
Please add defaults for these properties, or set them to `null` instead of `undefined`.

dblotsky commented on lib/ParamedicKill.js:
```diff
@@ -0,0 +1,95 @@
+    case util.WINDOWS:
+        return ["WWAHost.exe", "Xde.exe"];
+    case util.IOS:
+        return ["Simulator"];
```
Please add `"iOS Simulator"` to this array.

omefire replied:
consider returning an empty array if none of the platforms match this list.

dblotsky commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+var path 	 = require('path');
+var fs 		 = require('fs');
+var shelljs  = require("shelljs");
+var logger  = require('./utils').logger;
```
Nitpick: spacing.

dblotsky commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+
+ParamedicPermissions.prototype.updatePermissions = function(serviceList){
+	var self = this;
+	this.getDirectories(self.simulatorsFolder).forEach(function(simFolder){
```
If a regular for-loop is used here, the `self` pattern can be avoided.

dblotsky commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+			// No TCC.db file exists by default. So, Copy the new TCC.db file
+			var destinationTCCFolder = path.join(self.simulatorsFolder, simFolder, '/data/Library/TCC');
+			if(!self.doesFileExist(destinationTCCFolder)){
+				fs.mkdir(destinationTCCFolder, 0777);
```
Please use 0644 instead of 0777 here. Also, please factor out the permissions setting into a variable.

dblotsky commented on lib/paramedic.js:
```diff
@@ -105,10 +116,23 @@ ParamedicRunner.prototype.checkPlatformRequirements = function() {
 };
 
+ParamedicRunner.prototype.setPermissions = function() {
+    if(this.config.getPlatformId() === 'ios'){
```
Does this function only support iOS?

nikhilkh commented on lib/ParamedicLog.js:
```diff
@@ -0,0 +1,154 @@
+}
+
+ParamedicLog.prototype.logAndroid = function (){
+    var logCommand = "adb logcat -d";
```
Please use "adb logcat -d -v time"

nikhilkh commented on lib/ParamedicConfig.js:
```diff
@@ -53,7 +58,11 @@ ParamedicConfig.prototype.getReportSavePath = function () {
 
-ParamedicConfig.prototype.getShouldCleanUpAfterRun = function () {
+ParamedicConfig.prototype.setReportSavePath = function (reportSavePath) {
+    this._config.reportSavePath = reportSavePath;
```
Consider renaming this to `testResultsPath` or `testResultsDir` or `outputDir`

nikhilkh commented on lib/ParamedicConfig.js:
```diff
@@ -104,4 +117,44 @@ ParamedicConfig.prototype.getTimeout = function() {
+};
+
+ParamedicConfig.prototype.setLogPath = function(logPath) {
+    this._config.logPath = logPath;
```
Consider re-using `outputDir`

nikhilkh commented on lib/ParamedicConfig.js:
```diff
@@ -104,4 +117,44 @@ ParamedicConfig.prototype.getTimeout = function() {
+};
+
+ParamedicConfig.prototype.setAppName = function(appName) {
+    this._config.appName = appName;
```
Either remove this config parameter & use a default value that you can assume to be true 
or create the app with the name specified here.

nikhilkh commented on lib/ParamedicConfig.js:
```diff
@@ -104,4 +117,44 @@ ParamedicConfig.prototype.getTimeout = function() {
+};
+
+ParamedicConfig.prototype.setSimulatorsFolder = function(simulatorsFolder) {
+    this._config.simulatorsFolder = simulatorsFolder;
```
Consider using what we do for getting simulator logs. Ideally we should have one way of doing this and not configurable externally.

nikhilkh commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+}
+
+ParamedicPermissions.prototype.doesFileExist = function(filePath) {
+	var fileExists = false;
```
consider factoring it out into `util` module

nikhilkh commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+		//Check if the simFolder has TCC.db file in it
+		if(self.doesFileExist(destinationTCCFile)) {
+			var sqlite3Command_firstPart  = 'sqlite3 ' + destinationTCCFile;
+			var sqlite3Command_secondPart = ' "insert into access(service, client, client_type, allowed, prompt_count, csreq) values(\'';
```
Instead of copying a tcc db with permissions - consider copying empty and insert permissions into it - that helps in configuring permissions only in one place.

nikhilkh commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+			var sqlite3Command_secondPart = ' "insert into access(service, client, client_type, allowed, prompt_count, csreq) values(\'';
+			var sqlite3Command_lastPart   = '\',\'' + self.appName + '\',0,1,1,NULL)"';
+			serviceList.forEach(function(service){
+				var command = sqlite3Command_firstPart  +  sqlite3Command_secondPart + service + sqlite3Command_lastPart ;
```
Consider using `util.format`

nikhilkh commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+			}
+			var command = "cp " + self.dbPath + " " + destinationTCCFolder;
+			logger.info("Running Command: " + command);
+			shelljs.exec(command, {silent: true, async: false});
```
Consider writing a comment of why you are not handling errors here.

nikhilkh commented on lib/ParamedicLog.js:
```diff
@@ -0,0 +1,154 @@
+
+    var result = shelljs.exec(logCommand, {silent: true, async: false});
+    if (result.code > 0) {
+        logger.error("Failed to run command: " + logCommand);
```
Ideally you want to log return code and stderr/stdout

nikhilkh commented on lib/ParamedicLog.js:
```diff
@@ -0,0 +1,154 @@
+    return path.join(this.logPath, this.platform+"_logs.txt");
+}
+
+ParamedicLog.prototype.getLocalCLI = function (){
```
Consider just using "cordova" inline instead of this method.

nikhilkh commented on lib/ParamedicLog.js:
```diff
@@ -0,0 +1,154 @@
+}
+
+ParamedicLog.prototype.getLogFileName = function() {
+    return path.join(this.logPath, this.platform+"_logs.txt");
```
spaces missing

nikhilkh commented on lib/paramedic.js:
```diff
@@ -210,13 +234,29 @@ ParamedicRunner.prototype.waitForConnection = function() {
 
+ParamedicRunner.prototype.killProcess = function() {
+    if(this.config.shouldCleanUpAfterRun()){
+        logger.info("cordova-paramedic: Killing the process.");
```
emulator process?

rakatyal commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+	});
+}
+
+ParamedicPermissions.prototype.getDirectories = function(srcpath) {
```
Consider renaming it srcPath.

omefire commented on lib/ParamedicKill.js:
```diff
@@ -0,0 +1,95 @@
+    var platformTasks = this.tasksOnPlatform(this.platform);
+
+    if (platformTasks.length < 1) {
+        console.warn("no known tasks to kill");
```
in this case, we want to stop processing and bail out of the ```kill``` function.
Please, add a ```return``` statement here.

omefire commented on lib/ParamedicKill.js:
```diff
@@ -0,0 +1,95 @@
+        } else {
+            return ["emulator64-x86", "emulator64-arm"];
+        }
+        break;
```
this ```break``` statement will never be reached.

purplecabbage replied:
Personally I prefer NOT returning from a switch, and having a break for every condition ( or a commented fall through )

```
var retTaskList = [];
switch(platformName) {
  case util.WINDOWS :
      retTaskList = ["WWAHost.exe", "Xde.exe"];
      break;
 ...
} 
return retTaskList;
```

sgrebnov commented on main.js:
```diff
@@ -25,7 +25,12 @@ var USAGE = "Error missing args. \n" +
+    "--logMins: (optional) Windows only - specifies number of minutes to get logs\n" +
+    "--appName: (optional) iOS only - specifies the name of the app to be deployed\n" +
+    "--simulatorsFolder: (optional) iOS only - specifies the path for the simulatorsFolder\n" +
+    "--tccDbPath: (optional) iOS only - specifies the path for the TCC.db file to be copied.";
```
Readme.md should be updated as well

sgrebnov commented on lib/paramedic.js:
```diff
@@ -105,10 +116,23 @@ ParamedicRunner.prototype.checkPlatformRequirements = function() {
         throw new Error('Platform requirements check has failed!');
 };
 
+ParamedicRunner.prototype.setPermissions = function() {
```
Permission logic could be implemented via separate plugin with hooks

omefire commented on lib/ParamedicLog.js:
```diff
@@ -0,0 +1,154 @@
+    // out the path to that file, we need to find the ID of the simulator running
+    // mobilespec
+
+    // First, figure out the simulator that ran mobilespec. "cordova run"" just chooses
```
I find this to be very fragile code because it depends on internal implementation details of ```cordova run --list```. Those could well change in the future. I think it  might be better to just copy the logic that deals with listing the simulators in this project.
What do you think ?

omefire commented on lib/ParamedicPermissions.js:
```diff
@@ -0,0 +1,60 @@
+function ParamedicPermissions(appName, simulatorsFolder, dbPath) {
+	this.appName = appName;
+	this.simulatorsFolder = simulatorsFolder;
+	this.dbPath = dbPath;
```
nitpick: 'dbPath' -> 'tccDbPath'

sarangan12 commented:
@riknoll - Response for your comments
1. Why split this variable into three parts like this?
   Changed the logic to use util.format
2. Q is being required twice
   Removed it
3. Can paramedicPermissions be renamed to make it clear that it's only for iOS? The other naming convention seems like it applies to the generic files.
   Renamed it to ParamediciOSPermissions

@dblotsky - Response for your comments
1. Please add defaults for these properties, or set them to null instead of undefined.
   Changed the defaults to null.
2. Please add "iOS Simulator" to this array.
   Added
3. Nitpick: spacing. (in ParamedicPermissions.js (now ParamediciOSPermissions.js))
   Fixed it
4. If a regular for-loop is used here, the self pattern can be avoided.
   Removed it
5. Please use 0644 instead of 0777 here. Also, please factor out the permissions setting into a variable.
   Refactored the permission. But 0644 is not working. So keeping it as 0777.
6. Does this function only support iOS?
   Yes.

@nikhilkh - Response for your comments
1. Please use "adb logcat -d -v time"
   Done.
2. Consider renaming this to testResultsPath or testResultsDir or outputDir
   Done
3. Consider re-using outputDir
   Done
4. Either remove this config parameter & use a default value that you can assume to be true or create the app with the name specified here.
   Using a default value.
5. Consider using what we do for getting simulator logs. Ideally we should have one way of doing this and not configurable externally.
   Changed the logic and refactored the code between logs and permissions.
6. consider factoring it out into util module
   Done
7. Instead of copying a tcc db with permissions - consider copying empty and insert permissions into it - that helps in configuring permissions only in one place.
   Done
8. Consider using util.format
   Done.
9. Consider writing a comment of why you are not handling errors here.
   Done
10. Ideally you want to log return code and stderr/stdout
    Done
11. Consider just using "cordova" inline instead of this method.
    Done
12. spaces missing
    Fixed
13. emulator process?
    Fixed

@rakatyal - Response for your comments

1. Consider renaming it srcPath.
   Specific code has been removed

@omefire - Response for your comments
1. consider returning an empty array if none of the platforms match this list.
   Done.
2. in this case, we want to stop processing and bail out of the kill function. Please, add a return statement here.
   Done
3. this break statement will never be reached.
   Fixed it
4. nitpick: 'dbPath' -> 'tccDbPath'
   Fixed it.

@sgrebnov - Response for your comments
1. Readme.md should be updated as well
   Done
2. Permission logic could be implemented via separate plugin with hooks
   Hooks will be handled as a seperate user story in the following sprinits. I will create a task for this.

dblotsky commented on lib/ParamediciOSPermissions.js:
```diff
@@ -0,0 +1,43 @@
+var logger   = require('./utils').logger;
+var util     = require('./utils').utilities;
+
+var TCC_FOLDER_PERMISSION = 0777;
```
Why doesn't `0644` work? If execution permissions are required on the folder, then please use `0755`. Using `0777` is a security hazard.

sarangan12 replied:
Done

nikhilkh commented on lib/ParamedicConfig.js:
```diff
@@ -36,8 +36,10 @@ ParamedicConfig.parseFromArguments = function (argv) {
-        cleanUpAfterRun:   !!argv.cleanUpAfterRun? true: false
+        outputDir:         !!argv.outputDir? argv.outputDir: null,
+        cleanUpAfterRun:   !!argv.cleanUpAfterRun? true: false,
+        logMins:           !!argv.logMins? argv.logMins: null,
```
It might be a good idea to initialize defaults for optional parameters here e.g. util.DEFAULT_LOG_TIME

sarangan12 replied:
We are initializing it in paramedic.js

nikhilkh commented on lib/ParamedicConfig.js:
```diff
@@ -36,8 +36,10 @@ ParamedicConfig.parseFromArguments = function (argv) {
+        outputDir:         !!argv.outputDir? argv.outputDir: null,
+        cleanUpAfterRun:   !!argv.cleanUpAfterRun? true: false,
+        logMins:           !!argv.logMins? argv.logMins: null,
+        tccDbPath:         !!argv.tccDbPath? argv.tccDbPath: null
```
consider renaming tccDbPath to tccDb - it's simpler

nikhilkh commented on lib/ParamediciOSPermissions.js:
```diff
@@ -0,0 +1,43 @@
+        var command = "cp " + this.tccDbPath + " " + destinationTCCFolder;
+        logger.info("Running Command: " + command);
+        shelljs.exec(command, {silent: true, async: false});
+    }
```
You should consider using shelljs.cp instead?

dblotsky replied:
The `ncp` module is also an option.

sarangan12 replied:
DOne

nikhilkh commented on main.js:
```diff
@@ -37,6 +39,30 @@ if (pathToParamedicConfig || // --config
+        paramedicConfig.setPlugins(argv.plugin);
+    }
+
+    if(argv.outputDir) {
```
Do you assume the outputDir exists? Perhaps you validate somewhere else - should you instead validate here?

nikhilkh commented on main.js:
```diff
@@ -37,6 +39,30 @@ if (pathToParamedicConfig || // --config
+        paramedicConfig.setLogMins(argv.logMins);
+    }
+
+    if(argv.tccDbPath){
```
Do you assume the tccDbPath exists? Perhaps you validate somewhere else - should you instead validate here?

nikhilkh commented:
:+1:

omefire commented on lib/utils/utilities.js:
```diff
@@ -0,0 +1,123 @@
+    });
+
+    return simId;
+}
```
So, running this code, I face an issue where the simId returned is null.
Here's the output of various bash commands executed here:
- 
```
nstruments -s devices | grep ^iPhone
iPhone 4s (9.3) [93EC91E3-2679-49DC-AC6B-CAB279E462AF] (Simulator)
iPhone 5 (9.3) [830AC555-1986-4B48-82C7-898849A50541] (Simulator)
iPhone 5s (9.3) [17726EFF-F3B9-435E-8882-CC6ED589C213] (Simulator)
iPhone 6 (9.3) [3BDE39CD-7998-4959-B2D8-7C08C33B20B1] (Simulator)
iPhone 6 Plus (9.3) [7235ADB5-8730-4D0A-821F-5469D1ED47D7] (Simulator)
iPhone 6s (9.3) [32745196-42F6-4927-ADA0-16BACE15FC99] (Simulator)
iPhone 6s (9.3) + Apple Watch - 38mm (2.2) [6A868DF2-F36A-4AEE-9EC1-B956375934B9] (Simulator)
iPhone 6s Plus (9.3) [A949CB17-350F-4B28-AD64-425339ACEDBF] (Simulator)
iPhone 6s Plus (9.3) + Apple Watch - 42mm (2.2) [E0C9BC62-83AA-45C5-B411-FBCB557DE32B] (Simulator)
```

omefire replied:
So, turns out this regex breaks in Xcode 7.3, because of the ```(Simulator)``` added at the end, I still think this code is too fragile as it depends on specific Xcode output, which might change in the future. Maybe, we should go back to searching through all folders like before ? 

In the meantime, reajusting the regex to ```var simIdRegex = /^([a-zA-Z\d ]+) \(([\d.]+)\) \[([a-zA-Z\d\-]*)\].*$/;``` fixes this issue.

sarangan12 replied:
Done

dblotsky commented on lib/utils/utilities.js:
```diff
@@ -0,0 +1,123 @@
+    getSimId: getSimId,
+    doesFileExist: doesFileExist,
+    getSqlite3InsertionCommand: getSqlite3InsertionCommand
+};
```
The name of this file is awkward. Consider just putting this code in `lib/utils.js`.

sarangan12 replied:
This seems to be fine

sarangan12 commented:
Merge complete. Closing the PR
